lib/travis/api/app/middleware/opencensus.rb

```diff
@@ -1,50 +1,149 @@
-            unless event.payload[:cached] || event.payload[:name] == 'CACHE'
-              handle_notification_event event
-            end
+      end
+      
+      def self.handle_notification_event event
+        span_context = ::OpenCensus::Trace.span_context
+        if span_context
+          span = span_context.start_span event.name, skip_frames: 2
+          span.start_time = event.time
+          span.end_time = event.end
+          event.payload.each do |k, v|
+            span.put_attribute "#{k}", v.to_s
           end
         end
+      end
+      
+      ##
+      # Run the middleware.
+      #
+      # @param [Hash] env The rack environment
+      # @return [Array] The rack response. An array with 3 elements: the HTTP
+      #     response code, a Hash of the response headers, and the response
+      #     body which must respond to `each`.
+      #
+      def call env
+        formatter = AUTODETECTABLE_FORMATTERS.detect do |f|
+          env.key? f.rack_header_name
+        end
+        if formatter
+          context = formatter.deserialize env[formatter.rack_header_name]
+        end
+
+        if env['HTTP_TRACE'] == 'true' 
+          sampler = ::OpenCensus::Trace::Samplers::AlwaysSample.new
+        else 
+          sampler = ::OpenCensus::Trace.config.default_sampler
+        end
 
-        def handle_notification_event event
-          span_context = ::OpenCensus::Trace.span_context
-          if span_context
-            span = span_context.start_span event.name, skip_frames: 2
-            span.start_time = event.time
-            span.end_time = event.end
-            event.payload.each do |k, v|
-              span.put_attribute "#{k}", v.to_s
+        ::OpenCensus::Trace.start_request_trace \
+        trace_context: context,
+        same_process_as_parent: false do |span_context|
+          begin
+            span_context.in_span get_path(env), sampler: sampler do |span|
+              start_request span, env
+              @app.call(env).tap do |response|
+                finish_request span, response
+              end
             end
+          ensure
+            @exporter.export span_context.build_contained_spans
           end
         end
-        
+      end
+      
+      private
+      
+      def get_path env
+        path = "#{env['SCRIPT_NAME']}#{env['PATH_INFO']}"
+        path = "/#{path}" unless path.start_with? "/"
+        path
+      end
+      
+      def get_host env
+        env["HTTP_HOST"] || env["SERVER_NAME"]
+      end
+      
+      def get_url env
+        path = get_path env
+        host = get_host env
+        scheme = env["SERVER_PROTOCOL"]
+        query_string = env["QUERY_STRING"].to_s
+        url = "#{scheme}://#{host}#{path}"
+        url = "#{url}?#{query_string}" unless query_string.empty?
+        url
+      end
+      
+      def start_request span, env
+        span.kind = ::OpenCensus::Trace::SpanBuilder::SERVER
+        span.put_attribute "app", "travis-api"
```

Comment:
maybe change this to `api` to follow the convention we have in honeycomb and others?